File: app/src/test/java/com/dominikgaller/alpinebooking/architecture/ContextRegistryTest.java
package com.dominikgaller.alpinebooking.architecture;

import com.tngtech.archunit.core.domain.JavaClasses;
import com.tngtech.archunit.core.importer.ClassFileImporter;
import com.tngtech.archunit.core.importer.ImportOption;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Set;
import java.util.stream.Collectors;

import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * Mechanical enforcement of the bounded-context registry in
 * {@code documentation/architecture.definition.md} section 11.
 *
 * <p>The registry exists so that context boundaries are a checkable fact rather than a
 * matter of opinion. These tests make adding a fifth top-level package fail the build.
 *
 * <p>SDD: See {@code documentation/adr/0007-archunit-boundary-enforcement.adr.md} and
 *          {@code documentation/adr/0005-bounded-context-identity-boundaries.adr.md}.
 */
class ContextRegistryTest {

    private static final String ROOT = "com.dominikgaller.alpinebooking";

    /** The section 11 table. Adding a row here requires an ADR — and the ADR lands first. */
    private static final Set<String> REGISTERED_TOP_LEVEL_PACKAGES =
            Set.of("booking", "guide", "shared", "bootstrap");

    private static JavaClasses production;

    @BeforeAll
    static void importProductionClasses() {
        production = new ClassFileImporter()
                .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
                .withImportOption(location -> !location.contains("/jooq/"))
                .importPackages(ROOT);
    }

    /**
     * Guard against a silently empty import.
     *
     * <p>ArchUnit reads nothing at all if its bundled ASM does not understand the class
     * file version — which is exactly what happened on ArchUnit 1.3.0 and 1.4.0 against
     * the Java 25 baseline (ADR 0006): zero classes, no error. Every rule below would
     * then pass vacuously if ArchUnit lacked its own "failed to check any classes" guard.
     * This test makes the failure unmistakable rather than relying on that.
     */
    @Test
    @DisplayName("Guard: the importer actually found production classes")
    void importer_findsProductionClasses() {
        assertThat(production.size())
                .as("ArchUnit imported no classes — most likely its ASM cannot read the "
                        + "current class file version. See ADR 0006 and ADR 0007.")
                .isGreaterThan(50);
    }

    @Test
    @DisplayName("Section 11: top-level packages are exactly the registered ones")
    void topLevelPackages_areExactlyTheRegisteredSet() {
        final Set<String> onDisk = production.stream()
                .map(c -> c.getPackageName())
                .filter(p -> p.startsWith(ROOT + "."))
                .map(p -> p.substring(ROOT.length() + 1))
                .map(p -> p.contains(".") ? p.substring(0, p.indexOf('.')) : p)
                .collect(Collectors.toSet());

        assertThat(onDisk)
                .as("A top-level package not registered in architecture.definition.md "
                        + "section 11 is drift, however sensible it looks. A new bounded "
                        + "context is an ADR trigger (sdd.playbook.md section 6 item 9); "
                        + "adr/0003 is the precedent.")
                .isEqualTo(REGISTERED_TOP_LEVEL_PACKAGES);
    }

    /**
     * Section 11 rule 3 permits a context to depend on another's {@code core.inport} — its
     * published API — and nothing else. These two tests therefore forbid the *internals*
     * rather than the whole context.
     *
     * <p>The permitted form is a synchronous call from the orchestrating driver to the other
     * context's {@code *UseCase}, which is ordinary orchestration (§ 4.4). An outport
     * declared for the same purpose would only add an interface whose one implementation
     * delegates to that same inport.
     */
    @Test
    @DisplayName("Section 11 rule 3: booking does not import guide internals")
    void booking_doesNotImportGuideInternals() {
        noClasses()
                .that().resideInAPackage(ROOT + ".booking..")
                .should().dependOnClassesThat().resideInAnyPackage(
                        ROOT + ".guide.core.domain..",
                        ROOT + ".guide.core.outport..",
                        ROOT + ".guide.inbound..",
                        ROOT + ".guide.outbound..")
                .because("a context's core.inport is its published API and may be depended "
                        + "on; everything else is closed (section 11 rule 3).")
                .check(production);
    }

    @Test
    @DisplayName("Section 11 rule 3: guide does not import booking internals")
    void guide_doesNotImportBookingInternals() {
        noClasses()
                .that().resideInAPackage(ROOT + ".guide..")
                .should().dependOnClassesThat().resideInAnyPackage(
                        ROOT + ".booking.core.domain..",
                        ROOT + ".booking.core.outport..",
                        ROOT + ".booking.inbound..",
                        ROOT + ".booking.outbound..")
                .because("a context's core.inport is its published API and may be depended "
                        + "on; everything else is closed (section 11 rule 3).")
                .check(production);
    }

    /**
     * The permission is for the inport only, and only from a driver. A cross-context call
     * belongs in the layer whose job is orchestration — not in a controller, a listener or
     * an adapter, which would scatter the coupling across the delivery surface.
     */
    @Test
    @DisplayName("Section 11 rule 3: only a booking driver may reach guide's inport")
    void bookingReachesGuideInport_onlyFromADriver() {
        noClasses()
                .that().resideInAPackage(ROOT + ".booking..")
                .and().resideOutsideOfPackage(ROOT + ".booking.inbound.driver..")
                .should().dependOnClassesThat().resideInAPackage(ROOT + ".guide.core.inport..")
                .allowEmptyShould(true)
                .because("only inbound.driver orchestrates (section 4.4). A controller, "
                        + "listener or adapter reaching into another context spreads the "
                        + "coupling across the delivery surface. Depending on this context's "
                        + "own inport is of course fine and is not what this checks.")
                .check(production);
    }

    @Test
    @DisplayName("Section 11 rule 3: only a guide driver may reach booking's inport")
    void guideReachesBookingInport_onlyFromADriver() {
        noClasses()
                .that().resideInAPackage(ROOT + ".guide..")
                .and().resideOutsideOfPackage(ROOT + ".guide.inbound.driver..")
                .should().dependOnClassesThat().resideInAPackage(ROOT + ".booking.core.inport..")
                .allowEmptyShould(true)
                .because("only inbound.driver orchestrates (section 4.4). This is the rule "
                        + "UC12 relies on: CancelTourByGuideDriver may call booking's inport, "
                        + "GuideTourController may not.")
                .check(production);
    }

    @Test
    @DisplayName("Section 11 rule 4: shared depends on no bounded context")
    void shared_dependsOnNoBoundedContext() {
        noClasses()
                .that().resideInAPackage(ROOT + ".shared..")
                .should().dependOnClassesThat()
                .resideInAnyPackage(ROOT + ".booking..", ROOT + ".guide..")
                .because("architecture.definition.md section 9: any context may depend on "
                        + "shared; shared must not depend on any context.")
                .check(production);
    }

    @Test
    @DisplayName("Section 9: adapters for shared.outport live in shared.outbound")
    void sharedPortAdapters_liveInSharedOutbound() {
        noClasses()
                .that().resideInAnyPackage(ROOT + ".booking.outbound..", ROOT + ".guide.outbound..")
                .should().implement(ROOT + ".shared.outport.ClockPort")
                .orShould().implement(ROOT + ".shared.outport.DomainEventPublisher")
                .because("architecture.definition.md section 9: an implementation of a "
                        + "shared port belongs to no context, so it lives in "
                        + "shared.outbound and is wired by bootstrap.SharedConfig. Putting "
                        + "it in one context forces the other to obtain it from that "
                        + "context's configuration.")
                .check(production);
    }

    /**
     * ADR 0005: an identity owned by one context must not be imported by another. Foreign
     * identities cross the boundary as plain {@code String}.
     */
    @Test
    @DisplayName("ADR 0005: no context imports another context's identity type")
    void identities_doNotCrossContextBoundaries() {
        noClasses()
                .that().resideInAPackage(ROOT + ".booking..")
                .should().dependOnClassesThat()
                .haveFullyQualifiedName(ROOT + ".guide.core.domain.guidetour.GuideTourId")
                .because("ADR 0005: guideTourId crosses into booking as a plain String, so "
                        + "guide can change its identity representation without recompiling "
                        + "booking.")
                .check(production);
    }
}

File: src/test/kotlin/com/jobradleasing/contractmanagement/architecture/ContextRegistryTest.kt
package com.jobradleasing.contractmanagement.architecture

import com.tngtech.archunit.base.DescribedPredicate
import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.domain.JavaMethodCall
import com.tngtech.archunit.core.importer.ClassFileImporter
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

private const val BASE_PACKAGE = "com.jobradleasing.contractmanagement"

/**
 * Enforces `architecture.definition.md` § 11: the bounded-context registry is a
 * checkable fact, not a matter of opinion.
 *
 * SDD: See `documentation/adr/0007-archunit-boundary-enforcement.adr.md`.
 */
class ContextRegistryTest {
    private val importedClasses: JavaClasses =
        ClassFileImporter()
            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
            .importPackages(BASE_PACKAGE)

    /**
     * ADR 0007's consequences: a silent import failure would make every other
     * rule in this suite pass vacuously, which is more dangerous than no suite
     * at all. This is the tripwire.
     */
    @Test
    fun importer_findsProductionClasses() {
        assertThat(importedClasses.size).isGreaterThan(20)
    }

    @Test
    fun topLevelPackages_areRegistered() {
        val registered = setOf("bootstrap", "masterleasing", "individualleasing", "shared")
        val prefix = "$BASE_PACKAGE."

        val actual =
            importedClasses
                .map { it.packageName }
                .filter { it.startsWith(prefix) }
                .map { it.removePrefix(prefix).substringBefore('.') }
                .toSet()

        assertThat(actual).isSubsetOf(registered)
    }

    /**
     * `architecture.definition.md` § 9 rule 3 and § 11 rule 4: `shared` must not
     * depend on any bounded context. `coding-style.definition.md` § 6.2 names this
     * exact test as the enforcement behind the shared-kernel exemption
     * (`Money`/`EmployerId`/`LessorId` throwing `IllegalArgumentException` rather
     * than a domain exception, because `shared` structurally cannot reach one).
     */
    @Test
    fun shared_dependsOnNoBoundedContext() {
        noClasses()
            .that()
            .resideInAPackage("..shared..")
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage("..masterleasing..", "..individualleasing..")
            .check(importedClasses)
    }

    /**
     * `architecture.definition.md` § 11 rule 3: cross-context communication to
     * another context's inport is ordinary orchestration and belongs to the
     * orchestrating driver (§ 4.4), never to a controller or a listener reaching
     * in directly. `cancel-master-leasing-contract.inport.spec.md` § (UC04) and
     * `terminate-contracts-by-master-contract.inport.spec.md` (UC08) name this test
     * verbatim as their enforcement. Vacuous today — `individualleasing` does not
     * exist yet — and starts protecting the moment UC04's driver is built.
     */
    @Test
    fun masterLeasingReachesIndividualLeasingInport_onlyFromADriver() {
        noClasses()
            .that()
            .resideInAPackage("..masterleasing..")
            .and()
            .resideOutsideOfPackage("..masterleasing.inbound.driver..")
            .should()
            .dependOnClassesThat()
            .resideInAPackage("..individualleasing.core.inport..")
            .check(importedClasses)
    }

    /**
     * The symmetric case: `issue-individual-leasing-contract.inport.spec.md`
     * (UC05) and `read-leasing-terms.inport.spec.md` (UC09) name this test
     * verbatim. Vacuous today for the same reason as above.
     */
    @Test
    fun individualLeasingReachesMasterLeasingInport_onlyFromADriver() {
        noClasses()
            .that()
            .resideInAPackage("..individualleasing..")
            .and()
            .resideOutsideOfPackage("..individualleasing.inbound.driver..")
            .should()
            .dependOnClassesThat()
            .resideInAPackage("..masterleasing.core.inport..")
            .allowEmptyShould(true)
            .check(importedClasses)
    }

    /**
     * `architecture.definition.md` § 11 rule 3: everything except a context's
     * inport is closed to the other context. `read-leasing-terms.inport.spec.md`
     * names this test verbatim. Vacuous today for the same reason as above.
     */
    @Test
    fun individualLeasing_doesNotImportMasterLeasingInternals() {
        noClasses()
            .that()
            .resideInAPackage("..individualleasing..")
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage(
                "..masterleasing.core.domain..",
                "..masterleasing.core.outport..",
                "..masterleasing.inbound..",
                "..masterleasing.outbound..",
            ).allowEmptyShould(true)
            .check(importedClasses)
    }

    /**
     * `aggregate-master-leasing-contract.spec.md` § 4 (`reconstitute`):
     * "persistence mapper only ... must stay public because the mapper lives in
     * another package, so visibility cannot express the restriction". Named
     * verbatim there as this test's job. Matches by method name rather than by a
     * concrete target, so it protects from the day `reconstitute` is first written
     * — it does not need the method to exist yet to be a real (non-vacuous)
     * `noClasses` rule.
     */
    @Test
    fun reconstitute_isCalledOnlyByPersistenceMappers() {
        val callsReconstitute =
            object : DescribedPredicate<JavaMethodCall>("calls a method named 'reconstitute'") {
                override fun test(call: JavaMethodCall) = call.target.name == "reconstitute"
            }

        noClasses()
            .that()
            .resideOutsideOfPackage("..outbound.persistence..")
            .should()
            .callMethodWhere(callsReconstitute)
            .check(importedClasses)
    }
}
